dig.go

```diff
@@ -515,12 +552,18 @@ func (c *Container) provide(ctor interface{}, opts provideOptions) error {
 	}
 
 	for k := range keys {
-		oldProducers := c.providers[k]
-		c.providers[k] = append(oldProducers, n)
+		c.isVerifiedAcyclic = false
+		if c.deferAcyclicVerification {
+			c.providers[k] = append(c.providers[k], n)
+			continue
+		}
+		oldProviders := c.providers[k]
+		c.providers[k] = append(oldProviders, n)
 		if err := verifyAcyclic(c, n, k); err != nil {
-			c.providers[k] = oldProducers
+			c.providers[k] = oldProviders
```

Comment:
nit: I think we can avoid appending to providers twice since we're going to do
the first append unconditionally

```
for k := range keys {
    oldProviders := c.providers[k]
    c.providiers[k] = append(oldProviders, n)

    if c.deferAcyclicVerification {
        c.isVerifiedAcyclic = false
        continue
    }

    if err := verifyAcylic(c, n, k); err != nil {
        c.providers[k] = oldProviders
        return err
    }

    c.isVerifiedAcyclic = true
}
```